src/Psl/Regex/matching.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Regex;
+
+use Psl\Regex\Exception\RuntimeException;
+use Psl\Type\Exception\CoercionException;
+use Psl\Type\TypeInterface;
+
+use function preg_match;
+
+/**
+ * Determine if $subject matches the given $pattern and return the matches.
+ *
+ * @template T of array
+ *
+ * @param non-empty-string $pattern        The pattern to match against.
+ * @param TypeInterface<T> $capture_groups What shape does the matching items have?
+ *
+ * @return T
+ *
+ * @throws Exception\RuntimeException        If an internal error accord.
+ * @throws CoercionException                 If the capture groups don't match the pattern matches
+ * @throws Exception\InvalidPatternException If $pattern is invalid.
+ */
+function matching(string $subject, string $pattern, TypeInterface $capture_groups, int $offset = 0): array
+{
+    $matching = Internal\call_preg(
+        'preg_match',
+        static function () use ($subject, $pattern, $offset): array {
+            $matching = [];
+            $matches = preg_match($pattern, $subject, $matching, 0, $offset);
+
+            if ($matches === 0) {
+                throw new RuntimeException(
+                    'Could not match the pattern ' . $pattern . ' to the given input: ' . $subject . '.'
+                );
+            }
```

Comment:
how about `Psl\invariant(matches($subject, $pattern), '$subject must match $pattern');` instead? or nullable return?

Reply:
In this case, I don't like the invariant - since it throws a too generic exception.
Maybe better to introduce a `NoMatchesFound` exception instead of `RuntimeException`?


About nullable vs Exception : how do you want it to behave?
Most of the times you are expecting stuff to be there. So in that case an exception might make sense.
But maybe other problems lean more to a nullable solution.

Reply:
I think nullable makes sense in here, regex is expensive ( CPU wise ), so having to run `preg_matches` twice ( first to check if string matches, then to retrieve the matching parts ) is not ideal.

Reply:
This does mean that psalm makes it `null|array{...}` and you kinda always have to check for null on it.
+ it complicates the psalm plugin a bit (no worries about that)

Reply:
yep, but better than wasting CPU on regex.